### DISCUSSION TITLE
Docker runtime environment variables not possible

### DISCUSSION THREAD

**Discussion by BoKKeR** [vercel/next.js#39080] [Help]
I have searched around for a while but did not find a definitive answer.

How would I use runtime environmental variables defined by `-e NEXT_PUBLIC_BASE_URL=test` in the client side of NEXTJS

The things I tried without any success are the following
1. setting next.config.js -> env
2. setting next.config.js -> publicRuntimeConfig,
3. using @next/env -> getConfig()

```js
module.exports = withOffline({
	serverRuntimeConfig: {},
	publicRuntimeConfig: {
		NEXT_PUBLIC_BASE_URL: process.env.NEXT_PUBLIC_BASE_URL,
	},
	env: {
		NEXT_PUBLIC_BASE_URL: process.env.NEXT_PUBLIC_BASE_URL,
	},
	reactStrictMode: true,
})
```

```js
import React, { useEffect } from 'react'
import Head from 'next/head'
import getConfig from 'next/config'

const MyApp = ({}) => {
	useEffect(() => {
		console.log(process.env.NEXT_PUBLIC_BASE_URL)

		const { publicRuntimeConfig } = getConfig()
		console.log(publicRuntimeConfig.NEXT_PUBLIC_BASE_URL)
	}, [])

	return (
		<>
			<Head>
				<title>Thumba</title>
				<meta name="viewport" content="width=device-width, initial-scale=1.0" />
			</Head>
			<Component {...pageProps} />
		</>
	)
}

export default MyApp
```

This is the best writeup I found was here, yet I cant get it to work still. It works great if I let docker copy over the .env file during buildTime but that is not a great solution.

https://www.saltycrane.com/blog/2021/04/buildtime-vs-runtime-environment-variables-nextjs-docker/#setting-dynamic-runtime-environment-variables

I can see the variables on the server side but I cant get them to show up on the client side. I am running  "next": "^12.2.1"

**Comment by wwwdepot** [comment]
You need to provide build-time variable ([docker docs](https://docs.docker.com/engine/reference/commandline/build/#set-build-time-variables---build-arg)) so Next.js can inline those env var during `next build`.

```bash
docker build -t nextjs-docker --build-arg NEXT_PUBLIC_BASE_URL=http://blah.blah .
```

At production, this line (client-side code) will become static.

```
console.log(process.env.NEXT_PUBLIC_BASE_URL) -> console.log('http://blah.blah')
```

**Reply by BoKKeR** [reply]
Yes but this is not a solution, what if you are building an image for others to consume with their own env variables. I found workaround which involves re-building the nextjs app on each run.

**Reply by GabenGar** [reply]
Well there is no other way, there is no such thing as "process.env" on a client machine. For envs to be dynamic you have to implement the dynamic config yourself and that means a bunch of async logic involving server-specific http endpoints which also means all config consumers within the client code have to be async.

**Comment by airtonix** [comment]
@BoKKeR `process.env` inside your react is actually the [`webpack.DefinePlugin`](https://webpack.js.org/plugins/environment-plugin/), which is just a find/replace at buildtime.

If you want runtime variables, then you need to run NextJS in server mode and then use another method to get your flags and config into the frontend.

I'd personally just stop using `process.env` on both server and client.

I'd recommend both :
- an api endpoint within your nextjs server that computes any flags/config based on path parameters: `/api/config/company/:companyId/payroll`
- a server side service that provides all non contextual flags/config as an object that you inject into some kind of top level provider and disperse with a hook throughout your app.

This has interesting application, because now you can use `nconf` and some other runtime secrets layer like: https://www.npmjs.com/package/@vanderlaan/nconf-secrets to inject and provide a service layer to query these values at runtime!
